### DISCUSSION TITLE
How to use virtual threads with @Asynchronous @Inject ExecutorService executorService with Quarkus on JDK21?

### DISCUSSION THREAD

**Discussion by dhoffer** [quarkusio/quarkus#42448] [Q&A] [answered]
I would like to use virtual threads for blocking and/or I/O related operations where pre JDK virtual threads were platform threads.  How can I configure Quarkus to use virtual threads for the following cases:

`@Asynchronous

@Inject
ExecutorService executorService`

[a comment by an automated account is not shown]

**Comment by ozangunalp** [comment]
Where does the `@Asynchronous` annotation come from? (It doesn't look like MP Fault-Tolerance annotation)

If you are looking to schedule tasks on VTs you can [inject the virtual threads executor service](https://quarkus.io/guides/virtual-threads#inject-the-virtual-thread-executor)

**Comment by mschorsch** [comment, marked as answer]
@ozangunalp 
>  Where does the @Asynchronous annotation come from? (It doesn't look like MP Fault-Tolerance annotation)

https://jakarta.ee/specifications/platform/10/apidocs/jakarta/enterprise/concurrent/asynchronous

@dhoffer I don't think Quarkus supports `@Asynchronous` but you can inject a managed executor see https://quarkus.io/guides/context-propagation#usage-example-for-completionstage.

For Virtual Thread support see 
- https://quarkus.io/guides/virtual-threads
- https://quarkus.io/blog/virtual-thread-1/

**Reply by cescoffier** [reply]
Yes, use the executor and avoid @Asynchronous.

**Reply by dhoffer** [reply]
Yes we use @Asynchronous from the fault tolerent module when we need do do some async work, that works fine but its using a platform thread we want to use a virtual thread.  In the managed executor link, looks like I can use ManagedExecutor go use the Quarkus thread pool but how to specify if that is platform threads or virutal threads?  

In one of the other links I see:
@Inject
@VirtualThreads
ExecutorService vThreads;

So is that how to get a VT ExecutorService ?

I still don't see a VT alturnative for @Asynchronous

**Reply by cescoffier** [reply]
That's how you get the virtual thread executor.

I would need to check with @Ladicek when he is back from PTO, but I don't believe we will ever support @Asynchronous with virtual threads.

**Reply by dhoffer** [reply]
Bummer regarding  @asynchronous, but regarding @VirtualThreads will it fall back to platform threads if not available?  E.g. not JDK21 or later.

**Reply by cescoffier** [reply]
Yes, the executor does that.

**Reply by Ladicek** [reply]
That is correct, SmallRye Fault Tolerance as integrated into Quarkus offloads all blocking `@Asynchronous` method calls to the Quarkus main thread pool.

(Note that this `@Asynchronous` annotation comes from MicroProfile Fault Tolerance. To the best of my knowledge, Quarkus doesn't support Jakarta Concurrency at all.)

**Reply by dhoffer** [reply]
@Ladicek That is unfortunate.  Quarkus does support VT for at least REST serevices and for Executors it's a shame that Quarkus is not consistant and always support VT where it makes sense and @Asynchronous is definately where VT support makes sense.

**Reply by cescoffier** [reply]
IT would only make sense if you combine @Asynchronous and @RunOnVirtualThreads, otherwise it would be very dangerous.

**Comment by dhoffer** [comment]
Yes I was thinking the same thing, combine @asynchronous
<https://github.com/asynchronous> and @RunOnVirtualThreads no need for
something new/different.  To me this is a required feature of Quarkus as
why should one part/feature only use platform threads?

-David
